snapcraft/parts/plugins/colcon_plugin.py

```diff
@@ -129,31 +146,36 @@ def _get_workspace_activation_commands(self) -> List[str]:
         specific functionality.
         """
 
-        # There are a number of unbound vars, disable flag
-        # after saving current state to restore after.
-        return [
-            'state="$(set +o); set -$-"',
-            "set +u",
-            # If it exists, source the stage-snap underlay
-            'if [ -f "${CRAFT_PART_INSTALL}/opt/ros/${ROS_DISTRO}/local_setup.sh" ]; then',
-            'COLCON_CURRENT_PREFIX="{path}" . "{path}/local_setup.sh"'.format(
-                path="${CRAFT_PART_INSTALL}/opt/ros/${ROS_DISTRO}"
-            ),
-            "fi",
-            'if [ -f "${CRAFT_PART_INSTALL}/opt/ros/snap/local_setup.sh" ]; then',
-            'COLCON_CURRENT_PREFIX="{path}" . "{path}/local_setup.sh"'.format(
-                path="${CRAFT_PART_INSTALL}/opt/ros/snap"
-            ),
-            "fi",
-            '. "/opt/ros/${ROS_DISTRO}/local_setup.sh"',
-            'eval "${state}"',
-        ]
+        activation_commands = []
+
+        # Source ROS ws in all build-snaps first
+        activation_commands.append("## Sourcing ROS ws in build snaps")
+        self._options: ColconPluginProperties
+        if self._options.build_snaps:
+            for build_snap in self._options.build_snaps:
+                snap_name = _get_parsed_snap(build_snap)[0]
+                activation_commands.extend(
+                    self._get_source_command(f"/snap/{snap_name}/current")
+                )
+            activation_commands.append("")
+
+        # Source ROS ws in stage-snaps next
```

Comment:
Why would we need to source workspace from the stage-snaps? Stage-snaps are only meant to be used at run time right?

Reply:
Stage snap are another mechanism that isn't use in the context of this feature (content-sharing). Their use is already in the code base and I'm merely moving some code around here.